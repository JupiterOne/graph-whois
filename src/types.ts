import { IntegrationStepExecutionContext } from '@jupiterone/integration-sdk-core';

export type IntegrationStepContext = IntegrationStepExecutionContext<
  IntegrationConfig
>;

export interface IntegrationConfig {
  /**
   * Comma-delimited list of domain names
   *
   * (e.g. google.com,facebook.com)
   */
  domains: string;
}
