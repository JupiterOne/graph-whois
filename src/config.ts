import {
  IntegrationExecutionContext,
  IntegrationInstanceConfig,
  IntegrationInstanceConfigFieldMap,
  IntegrationValidationError,
} from '@jupiterone/integration-sdk-core';
import isValidDomain from 'is-valid-domain';
import { getDomainListFromConfigParam } from './utils';

export const instanceConfigFields: IntegrationInstanceConfigFieldMap = {
  domains: {
    type: 'string',
    mask: false,
  },
};

export interface IntegrationConfig extends IntegrationInstanceConfig {
  /**
   * Comma-delimited list of domain names
   *
   * (e.g. google.com,facebook.com)
   */
  domains: string;
}

export async function validateInvocation(
  context: IntegrationExecutionContext<IntegrationConfig>,
) {
  const {
    instance: { config },
  } = context;

  if (!config.domains) {
    throw new IntegrationValidationError(
      'config.domains must be provided by the user',
    );
  }

  const domains = getDomainListFromConfigParam(config.domains);

  for (const domain of domains) {
    if (!isValidDomain(domain)) {
      throw new IntegrationValidationError(`Invalid domain name ${domain}`);
    }
  }

  return Promise.resolve();
}
