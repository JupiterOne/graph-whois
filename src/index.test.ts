import {
  Recording,
  createMockStepExecutionContext,
  setupRecording,
} from '@jupiterone/integration-sdk-testing';
import { IntegrationConfig } from './types';
import {
  ENTITY_DOMAIN_CLASS,
  ENTITY_DOMAIN_TYPE,
  fetchDomains,
  createDomainEntity,
} from '.';
import { getMockIntegrationConfig } from '../test/config';

describe('#fetchDomains', () => {
  let recording: Recording;

  beforeEach(() => {
    recording = setupRecording({
      directory: __dirname,
      name: 'fetchDomains',
    });
  });

  afterEach(async () => {
    await recording.stop();
  });

  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: getMockIntegrationConfig(),
    });

    await fetchDomains(context);

    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      numCollectedRelationships: context.jobState.collectedRelationships.length,
      // Since fetching domain details happens in parallel, we need to sort the
      // entities alphabetically by _key, so that the snapshot results remain
      // in the correct order.
      collectedEntities: context.jobState.collectedEntities.sort((a, b) => {
        if (a._key > b._key) return 1;
        if (a._key < b._key) return -1;
        return 0;
      }),
      collectedRelationships: context.jobState.collectedRelationships,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();

    expect(context.jobState.collectedEntities).toMatchGraphObjectSchema({
      _class: ENTITY_DOMAIN_CLASS,
      schema: {
        additionalProperties: false,
        properties: {
          _type: { const: ENTITY_DOMAIN_TYPE },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          domainName: { type: 'string' },
          expiresOn: { type: 'number' },
          whoisServer: { type: 'string' },
          registrar: { type: 'string' },
          status: { type: 'string' },
        },
      },
    });
  });
});

describe('#createDomainEntity', () => {
  test('should convert to entity', () => {
    expect(
      createDomainEntity({
        id: '2138514_DOMAIN_COM-VRSN',
        whois_server: 'whois.markmonitor.com',
        updated_date: '2019-09-09T15:39:04Z',
        creation_date: '1997-09-15T04:00:00Z',
        expiration_date: '2028-09-14T04:00:00Z',
        registrar: 'MarkMonitor Inc.',
        emails: 'abusecomplaints@markmonitor.com',
        status:
          'serverUpdateProhibited https://icann.org/epp#serverUpdateProhibited',
        nameservers: 'NS4.GOOGLE.COM',
        contact: {
          registrant: {},
          technical: {},
          admin: {},
          billing: {},
        },
        name: 'google.com',
      }),
    ).toMatchSnapshot();
  });
});
