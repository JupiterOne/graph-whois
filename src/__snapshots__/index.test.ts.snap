// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#createDomainEntity should convert to entity 1`] = `
Object {
  "_class": Array [
    "Domain",
  ],
  "_key": "internet_domain_google.com",
  "_rawData": Array [
    Object {
      "name": "default",
      "rawData": Object {
        "contact": Object {
          "admin": Object {},
          "billing": Object {},
          "registrant": Object {},
          "technical": Object {},
        },
        "creation_date": "1997-09-15T04:00:00Z",
        "emails": "abusecomplaints@markmonitor.com",
        "expiration_date": "2028-09-14T04:00:00Z",
        "id": "2138514_DOMAIN_COM-VRSN",
        "name": "google.com",
        "nameservers": "NS4.GOOGLE.COM",
        "registrar": "MarkMonitor Inc.",
        "status": "serverUpdateProhibited https://icann.org/epp#serverUpdateProhibited",
        "updated_date": "2019-09-09T15:39:04Z",
        "whois_server": "whois.markmonitor.com",
      },
    },
  ],
  "_type": "internet_domain",
  "active": undefined,
  "contactEmails": Array [
    "abusecomplaints@markmonitor.com",
  ],
  "createdOn": 874296000000,
  "displayName": "google.com",
  "domainName": "google.com",
  "expiresOn": 1852516800000,
  "id": "2138514_DOMAIN_COM-VRSN",
  "name": "google.com",
  "nameservers": Array [
    "NS4.GOOGLE.COM",
  ],
  "registrar": "MarkMonitor Inc.",
  "status": "serverUpdateProhibited https://icann.org/epp#serverUpdateProhibited",
  "updatedOn": 1568043544000,
  "whoisServer": "whois.markmonitor.com",
}
`;

exports[`#fetchDomains should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Domain",
      ],
      "_key": "internet_domain_facebook.com",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "contact": Object {
              "admin": Object {},
              "billing": Object {},
              "registrant": Object {},
              "technical": Object {},
            },
            "creation_date": "1997-03-29T05:00:00Z",
            "emails": "abusecomplaints@registrarsafe.com",
            "expiration_date": "2028-03-30T04:00:00Z",
            "id": "2320948_DOMAIN_COM-VRSN",
            "name": "facebook.com",
            "nameservers": "D.NS.FACEBOOK.COM",
            "registrar": "RegistrarSafe, LLC",
            "status": "serverUpdateProhibited https://icann.org/epp#serverUpdateProhibited",
            "updated_date": "2020-03-10T18:53:59Z",
            "whois_server": "whois.registrarsafe.com",
          },
        },
      ],
      "_type": "internet_domain",
      "active": undefined,
      "contactEmails": Array [
        "abusecomplaints@registrarsafe.com",
      ],
      "createdOn": 859611600000,
      "displayName": "facebook.com",
      "domainName": "facebook.com",
      "expiresOn": 1838001600000,
      "id": "2320948_DOMAIN_COM-VRSN",
      "name": "facebook.com",
      "nameservers": Array [
        "D.NS.FACEBOOK.COM",
      ],
      "registrar": "RegistrarSafe, LLC",
      "status": "serverUpdateProhibited https://icann.org/epp#serverUpdateProhibited",
      "updatedOn": 1583866439000,
      "whoisServer": "whois.registrarsafe.com",
    },
    Object {
      "_class": Array [
        "Domain",
      ],
      "_key": "internet_domain_google.com",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "contact": Object {
              "admin": Object {},
              "billing": Object {},
              "registrant": Object {},
              "technical": Object {},
            },
            "creation_date": "1997-09-15T04:00:00Z",
            "emails": "abusecomplaints@markmonitor.com",
            "expiration_date": "2028-09-14T04:00:00Z",
            "id": "2138514_DOMAIN_COM-VRSN",
            "name": "google.com",
            "nameservers": "NS4.GOOGLE.COM",
            "registrar": "MarkMonitor Inc.",
            "status": "serverUpdateProhibited https://icann.org/epp#serverUpdateProhibited",
            "updated_date": "2019-09-09T15:39:04Z",
            "whois_server": "whois.markmonitor.com",
          },
        },
      ],
      "_type": "internet_domain",
      "active": undefined,
      "contactEmails": Array [
        "abusecomplaints@markmonitor.com",
      ],
      "createdOn": 874296000000,
      "displayName": "google.com",
      "domainName": "google.com",
      "expiresOn": 1852516800000,
      "id": "2138514_DOMAIN_COM-VRSN",
      "name": "google.com",
      "nameservers": Array [
        "NS4.GOOGLE.COM",
      ],
      "registrar": "MarkMonitor Inc.",
      "status": "serverUpdateProhibited https://icann.org/epp#serverUpdateProhibited",
      "updatedOn": 1568043544000,
      "whoisServer": "whois.markmonitor.com",
    },
  ],
  "collectedRelationships": Array [],
  "encounteredTypes": Array [
    "internet_domain",
  ],
  "numCollectedEntities": 2,
  "numCollectedRelationships": 0,
}
`;
